### ISSUE TITLE
Consistent and automatic code formatting

### ISSUE THREAD

**Issue by drewnoakes** [gitextensions/gitextensions#4534]
In #4507 and other issues/PRs there's been discussion about formatting changes.

The code currently uses different formatting styles throughout, often within the same method.

There exist tools to automatically and consistently format source code.

Given the amount of PR feedback relating to code formatting, there would be less friction to contributing, both in terms of reading existing code and feedback on non-functional aspects of PRs.

Options for auto-formatting include:

- ReSharper (commercial, but quite common)
- [StyleCop Analysers](https://github.com/DotNetAnalyzers/StyleCopAnalyzers)
- others?

Then there is the set of rules themselves. People often have very different ideas about this as it is very subjective and it can be quite polarising. Hopefully the benefit of consistency and clarity will outweigh any personal preferences on specific points.

The style should also be documented in some form that can be referenced from `CONTRIBUTING.md`.

**Comment by crypto-rsa** [comment]
Another option would be to expand the existing [.editorconfig](https://github.com/gitextensions/gitextensions/blob/master/.editorconfig) file according to the [reference](https://docs.microsoft.com/en-gb/visualstudio/ide/editorconfig-code-style-settings-reference).

**Comment by pmiossec** [comment]
>others?

I was very pleased with code formatter: https://github.com/dotnet/codeformatter
the nuget package: https://www.nuget.org/packages/Octokit.CodeFormatter/1.0.0-preview
the command line: `".\packages\Octokit.CodeFormatter\tools\CodeFormatter.exe" .\GitExtensions.sln /rule-:FieldNames /nounicode /nocopyright`

I took some minutes to run it on GitExtensions and it fixed a lot of little inconsistencies. I will make a pull request for you to have a look.

But there is a at least one problem. That break the indentation of C#7 local functions 😭 on 10 files (I only commited the files well fixed)

**Comment by RussKie** [comment]
Re# is good but I don't think it can address all stylistic requirements. Also sadly not all devs enjoy the luxury of having it.

I am probably most interested in StyleCop analysers - if they prove good then I can borrow the approach for our work environment :grinning:

**Comment by drewnoakes** [comment]
Pushed up an example of what using StyleCop might be like.

We can make R# rules that match the StyleCop rules, and add them to the `.DotSettings` file(s) so that R# generates/reformats in a way that StyleCop is happy with.

**Comment by sharwell** [comment]
> if they prove good

They are outstanding, but only configurable to a limited extent. The bigger question you'll end up asking is how close it can come to your style preferences.

**Comment by sharwell** [comment]
> We can make R# rules that match the StyleCop rules, and add them to the .DotSettings file(s) so that R# generates/reformats in a way that StyleCop is happy with.

I'm pretty sure ReSharper can read the editorconfig settings, so if we add the rules there it should work for both users with and without ReSharper.

**Comment by vbjay** [comment]
I use coderush for roslyn, myself.

**Comment by drewnoakes** [comment]
Based on experience from #4539 and @sharwell's advice therein, I would like to propose a route forward that uses `StyleCop.Analyzer` (SCA).

1. Create a new PR that:
    - adds the `StyleCop.Analyzer` package to every project
    - create a new `GitExtensions.ruleset` that disables _all_ SCA rules 
    - configures every project in the solution to use this new ruleset
    - doesn't change any `.cs` files
2. Once that is merged, we create new PRs (likely, one-after-another) that enable single rules (or groups of closely related rules) across the whole solution, along with the corresponding code fixes.
    - As @sharwell suggests, we would commit the output of the automated fix in one commit, along with any manual tidy up required as a secondary commit (if necessary).

This approach will produce more churn and take longer to complete, but it should be more reviewable than going project-by-project. #4539 kicked the tyres on a single project (albeit one of the larger ones) and the diff was huge.

The experience of using SCA was quite nice. I was impressed. Violations fail the build, and there's a fair degree of control over the rules. Not as flexible as R#'s style options, but I don't think that matters in practice as the main style rules are covered. The automated fixes worked quite well and were available for the majority of the rules. In all, I lean towards recommending SCA as a solid answer to this issue.

I'll put together that first PR now as a starting point, as it shouldn't take too long. I'd be interested to hear any other feedback on this issue/approach.

**Comment by sharwell** [comment]
> create a new GitExtensions.ruleset ...

Create two rulesets: one for production code and one for test code. The test code should inherit the production code rules and provides a place to override behavior. Here is an example:

https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers.Internal.ruleset

For this project, the "Internal" ruleset would be empty except for the `<Include Path="GitExtensions.ruleset" Action="Default" />` line.

> ... that disables all SCA rules

This block will work:
https://github.com/sharwell/roslyn/blob/5d2a4e093878c7cee0a328106b71bda57cdccda4/build/Rulesets/Roslyn_BuildRules.ruleset#L96-L262

> This approach will produce more churn and take longer to complete ...

It's not too bad. Some of the rules have more churn, but you'll find many rules were either already followed exactly or were close to followed.

**Comment by drewnoakes** [comment]
@sharwell I attempted to capture your suggestions in #4547. Please take a look.

I had to add additional exclusions for SA0001 and SA1633 to make things compile.

**Comment by sharwell** [comment]
> I had to add additional exclusions for SA0001 and SA1633 to make things compile.

I meant to link you to a page that had SA1633 excluded. SA0001 makes sense, but you should make enabling that rule the very first change after the initial integration.

**Comment by drewnoakes** [comment]
@sharwell #4551 enables SA0001.

**Comment by pmiossec** [comment]
I remembered it perhaps too late but there is R# code cleanup (free command line tool) 
https://blog.jetbrains.com/dotnet/2018/03/01/code-cleanup-resharper-command-line-tools/

**Comment by RussKie** [comment]
Can we consider this done?

**Comment by drewnoakes** [comment]
Here's what's outstanding. We can choose to fix or ignore these, but let's at least discuss them.

### Member ordering

SCA has rules in the ([SA12xx family](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/OrderingRules.md)) related to member ordering. Here are the ones we have not yet enabled:

Identifier | Name | Description
-----------|------|-------------
SA1201 | ElementsMustAppearInTheCorrectOrder | An element within a C# code file is out of order in relation to the other elements in the code. 
SA1202 | ElementsMustBeOrderedByAccess | An element within a C# code file is out of order within regard to access level, in relation to other elements in the code.
SA1203 | ConstantsMustAppearBeforeFields | A constant field is placed beneath a non-constant field. 
SA1204 | StaticElementsMustAppearBeforeInstanceElements | A static element is positioned beneath an instance element of the same type. 
SA1214 | ReadonlyElementsMustAppearBeforeNonReadonlyElements | A readonly field is positioned beneath a non-readonly field. 

I personally think we could enable all of these except SA1202.

This would result in a consistent ordering across files having:

- constant fields
- static readonly fields
- static non-readonly fields
- instance readonly fields
- instance non-readonly fields
- other members

I haven't tried this out yet but can give it a shot and we can discuss.


### Naming rules

SCA has rules in the ([SA13xx family](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/NamingRules.md)) related to member naming. Here are the ones we have not yet enabled:

Identifier | Name | Description
-----------|------|------------
SA1300 | ElementMustBeginWithUpperCaseLetter | The name of a C# element does not begin with an upper-case letter. 
SA1303 | ConstFieldNamesMustBeginWithUpperCaseLetter | The name of a constant C# field should begin with an upper-case letter. 
SA1304 | NonPrivateReadonlyFieldsMustBeginWithUpperCaseLetter | The name of a non-private readonly C# field should being with an upper-case letter. 
SA1306 | FieldNamesMustBeginWithLowerCaseLetter | The name of a field in C# does not begin with a lower-case letter. 
SA1307 | AccessibleFieldsMustBeginWithUpperCaseLetter | The name of a public or internal field in C# does not begin with an upper-case letter. 
SA1310 | FieldNamesMustNotContainUnderscore | A field name in C# contains an underscore. 
SA1311 | StaticReadonlyFieldsMustBeginWithUpperCaseLetter | The name of a static readonly field does not begin with an upper-case letter. 

I tried SA1300 but there are so many event handlers with names like `button_Click` that it didn't feel like a good idea. Further, any new event handlers added via the WinForms designer would be violations. The underscores also trigger SA1310, which is why that rule's disabled too.

But some of the others seem sensible. We currently use `_lowerCase` for private fields, but maybe we use `PascalCase` for all other fields makes sense. If so, we could enable SA1303, SA1304, SA1307 and SA1311.

Again, we could try this out. I don't think that selection of rules would involve too much churn.

**Comment by sharwell** [comment]
> I personally think we could enable all of these except SA1202.

This is a reasonable balance, but overall this change can be particularly difficult to review since diff tools don't handle reordering with any elegance.

**Comment by gerhardol** [comment]
Whatever is done, make it quick. 
Suggestions seem reasonable.

Great job, much appreciated but I still would like to get down to normal business...

**Comment by pmiossec** [comment]
>[...]this change can be particularly difficult to review since diff tools don't handle reordering with any elegance.

That's not the subject here but I can't let the occasion pass... 

Git is now being able to do it since v2.15 with `--color-moved` option (or setting `git config --global diff.colorMoved`).

I wonder if it would be possible to support it easily in GitExtensions. From what I saw, no, because that's GitExtensions that do the coloring using the +/- in the first column of the diff.

But perhaps someone know better than me...

I think it could be great to have this feature.

(perhaps I should create an issue)

**Comment by RussKie** [comment]
I like all of the 12XX :) 
On a serious note, I'm happy not to turn 1202 now.

**Comment by RussKie** [comment]
* SA1300 - this will conflict with event handlers, so I don't think we should enable it
* SA1303 - good
* SA1304 - ok. @sharwell is there a rule to ban public fields? We aren't [Rico Mariani to break rules](https://blogs.msdn.microsoft.com/ricom/2006/08/31/performance-quiz-11-ten-questions-on-value-based-programming/)
* SA1306 - This will cause noise for `_NO_TRANSLATE_xxx` fields. I expected it to complain about the leading underscore, but it skips it.
* SA1307 - good
* SA1310 - will cause too much noise
* SA1311 - good, but we'll need to add a bunch of ignores for `GitVersion` object.

**Comment by sharwell** [comment]
> is there a rule to ban public fields?

You have [CA1051: Do not declare visible instance fields](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1051-do-not-declare-visible-instance-fields) for that, but mostly I believe it's up to code review.

**Comment by RussKie** [comment]
I believe we have addressed this for the most part, if not entirely
